File: DigitalCircuit/lib/signal.ts
function xor(a: any, b: any) {
    return !!a !== !!b;
}
export function signal(input:Boolean[],last:Boolean[]):Boolean[]{
	let output:Boolean[]=new Array(8).fill(false);
	//output[7,5,3,1]记录了本次的输入
	output[7]=input[3];
	output[5]=input[2];
	output[3]=input[1];
	output[1]=input[0];
	//output[6]只有在「本次输入与上次输入完全相反」时，才会得到true
	output[6]=xor(input[3],last[7])&&xor(input[2],last[5])&&xor(input[1],last[3])&&xor(input[0],last[1]);
	//output[4,2]组成二位计数器
	output[4]=xor(last[4],last[2]);
	output[2]=!last[2];
	//result[0]等于所有输入、（上次）输出的异或和
	output[0]=xor(input[0],xor(input[1],xor(input[2],xor(input[3],xor(last[0],xor(last[1],xor(last[2],xor(last[3],xor(last[4],xor(last[5],xor(last[6],last[7])))))))))));
	return output;
}

File: DigitalCircuit/lib/signal.cpp
#include<array>
#include"operate.hpp"
#include"preset.hpp"
std::array<bool,preset::OutputNum> operate::signal(std::array<bool,preset::InputNum> &input,std::array<bool,preset::OutputNum> &output){
	std::array<bool,preset::OutputNum> result;
	//result[7,5,3,1]记录了本次的输入
	result[7]=input[3];
	result[5]=input[2];
	result[3]=input[1];
	result[1]=input[0];
	//result[6]只有在「本次输入与上次输入完全相反」时，才会得到true
	result[6]=input[3]^output[7]&&input[2]^output[5]&&input[1]^output[3]&&input[0]^output[1];
	//result[4,2]组成二位计数器
	result[4]=output[4]^output[2];
	result[2]=!output[2];
	//result[0]等于所有输入、输出的异或和
	result[0]=false;
	for(int i=0;i<preset::InputNum;i++)
		result[0]^=input[i];
	for(int i=0;i<preset::OutputNum;i++)
		result[0]^=output[i];
	return result;
}
